iXBRLViewerPlugin/xhtmlserialize.py

```diff
@@ -13,29 +13,134 @@
+        if prefix is None:
+            return qname.localname
+        return "%s:%s" % (prefix, qname.localname)
+
+    def qname_for_attr(self, tag, nsmap):
+        qname = etree.QName(tag)
+        if qname.namespace is None:
+            return qname.localname
+        prefix = next(iter(sorted((p for p, ns in nsmap.items() if ns == qname.namespace and p is not None), key = self.prefix_sort)))
+        return "%s:%s" % (prefix, qname.localname)
+
+    def is_selfclosable(self, n):
+        qname = etree.QName(n)
+        return (
+                qname.localname in XHTMLSerializer.SELF_CLOSABLE
+                and (
+                    qname.namespace == XHTML_NS
+                    or (self.assume_xhtml and qname.namespace is None)))
+
+    def escape_attr(self, s):
+        return re.sub(r'([<>"&])', lambda m: self.ENTITIES[m.group(0)], s)
+
+    def xmlns_declaration(self, prefix, uri):
+        if prefix is None:
+            return 'xmlns="%s"' % self.escape_attr(uri)
+        return 'xmlns:%s="%s"' % (prefix, self.escape_attr(uri))
+
+    def namespace_declarations(self, new_nsmap, cur_nsmap):
+        changed = set(p[0] for p in (set(new_nsmap.items()) ^ set(cur_nsmap.items())))
+        return sorted(self.xmlns_declaration(p, new_nsmap[p]) for p in changed)
+
+    def escape_text(self, s, escape_mode):
+        if s is None:
+            return ''
+        if escape_mode == EscapeMode.STYLE:
+            # Don't escape >
+            # & and < may only appear in a string or comment in CSS, so escape
+            # as if they're in a string.
+            return re.sub(r'([<&])', lambda m: "\\%06X" % ord(m.group(1)), s)
+        return re.sub(r'([<>&])', lambda m: self.ENTITIES[m.group(0)], s)
+
+    def attributes(self, node):
+        return sorted(
+            '%s="%s"' % (self.qname_for_attr(k, node.nsmap), self.escape_attr(v))
+                for k, v in node.items()
+            )
+
+    def write_comment(self, n, escape_mode):
+        self.write('<!--' + n.text + '-->' + self.escape_text(n.tail, escape_mode))
+
+    def write_processing_instruction(self, n, escape_mode):
+        self.write( '<?' + n.target + ' ' + n.text + '?>' + self.escape_text(n.tail, escape_mode))
+
+    def write_element(self, n, parent_nsmap = {}, escape_mode = EscapeMode.DEFAULT):
+        name = self.qname_for_node(n)
+        qname = etree.QName(n.tag)
+        selfclose = len(n) == 0 and n.text is None and self.is_selfclosable(n)
+        parts = [ name ] + self.namespace_declarations(n.nsmap, parent_nsmap) + self.attributes(n)
+        self.write('<' + ' '.join(parts))
+
+        if qname.localname == 'style':
+            inner_escape_mode = EscapeMode.STYLE
+        else:
+            inner_escape_mode = escape_mode
+
+        if selfclose:
+            self.write('/>')
+        else:
+            self.write('>' + self.escape_text(n.text, inner_escape_mode))
```

Comment:
It might be worth streaming the `escape_text` call too: we have seen enormous `<style>` with embedded fonts, background images, etc.  It might even be nice to avoid the `.replace` call and manually splitting the string and writing it out in chunks.

@derekgengenbacher-wf It's probably worth testing this on some very large documents (perhaps SEC, ESEF, and FERC?) to verify that this doesn't cause any perf problems with time/memory or any rendering issues.

Reply:
Done in 1fd8e955af60e8d24e9056053042d931674e5702

I've left the `re.sub` in.  Testing with a fairly large ESEF document, it adds around 0.2s to an overall write time of 5s compared to writing the string out unmodified, so that's the upper bound of what we could save here.

Naive attempts to reimplement it in a streaming manner are significantly slower.  `re.sub` is implemented in C, so I think it may be hard to improve on in Python.